lnst/RecipeCommon/Perf/Measurements/ForwardingMeasurement.py

```diff
@@ -0,0 +1,469 @@
+
+        return job
+
+    def _prepare_client(self):
+        if not all(flow.generator == self.flows[0].generator for flow in self.flows):
+            raise MeasurementError("Multiple generators are not supported")
+
+        config = []
+        for flow in self.flows:
+            config.append(
+                {
+                    "src_if": self._real_dev(flow.generator_nic),
+                    "dst_mac": flow.forwarder_nic.hwaddr,
+                    "src_ip": flow.generator_bind,
+                    "dst_ip": flow.receiver_bind,
+                    "cpu": flow.generator_cpupin[
+                        0
+                    ],  # FwdMeasGen round-robins cpus, so this will be list with 1 cpu only
+                    "pkt_size": flow.msg_size,
+                    "duration": flow.duration + flow.warmup_duration * 2,
+                    "src_port": flow.generator_port,
+                    "dst_port": flow.receiver_port,
+                    "ratep": int(
+                        self._ratep / self._burst
+                    ),  # pktgen internally does ratep * burst
+                    # ^ ratep should be set, to prevent bandwidth starvation
+                    "burst": self._burst,
+                }
+            )
+
+        pktgen = PktgenController(config=config)
+
+        job = self.flows[0].generator.prepare_job(pktgen)
+
+        return job
+
+    def _prepare_jobs(self):
+        self._client_job = self._prepare_client()
+        self._forwarder_job = self._prepare_forwarder()
+        self._server_job = self._prepare_server()
+
+        for flow in self.flows:
+            net_flow = NetworkFlowTest(flow, self._server_job, self._client_job)
+            net_flow.forwarder_job = self._forwarder_job
+
+            self._net_flows.append(net_flow)
+
+    def finish(self):
+        try:
+            self._client_job.wait(timeout=self._client_job.what.runtime_estimate())
+            self._forwarder_job.wait(
+                timeout=self._forwarder_job.what.runtime_estimate()
+            )
+            self._server_job.wait(timeout=self._server_job.what.runtime_estimate())
+        finally:
+            self._client_job.kill()
+            self._forwarder_job.kill()
+            self._server_job.kill()
+
+        self._finished_client_job = self._client_job
+        self._finished_server_job = self._server_job
+        self._finished_forwarder_job = self._forwarder_job
+
+        self._client_job = None
+        self._server_job = None
+        self._forwarder_job = None
+
+    def collect_results(self):
+        results = []
+
+        receiver_results = self._parse_receiver_results()
+        forwarder_results = self._parse_forwarder_results()
+        generator_results = self._parse_generator_results()
+
+        for net_flow in self._net_flows:
+            flow = net_flow.flow
+            result = ForwardingMeasurementResults(
+                measurement=self,
+                measurement_success=True,
```

Comment:
this should likely be calculated based on job status - if it was killed or otherwise crashed it is unlikely that the measurement is successful.